exp/services/ledgerexporter/internal/uploader.go

```diff
@@ -0,0 +1,72 @@
+package exporter
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+// Uploader is responsible for uploading data to a storage destination.
+type Uploader interface {
+	Run(ctx context.Context) error
+	Upload(ctx context.Context, metaArchive *LedgerMetaArchive) error
+}
+
+type uploader struct {
+	destination   DataStore
+	metaArchiveCh chan *LedgerMetaArchive
+}
+
+func NewUploader(destination DataStore, metaArchiveCh chan *LedgerMetaArchive) Uploader {
+	return &uploader{
+		destination:   destination,
+		metaArchiveCh: metaArchiveCh,
+	}
+}
+
+// Upload uploads the serialized binary data of ledger TxMeta to the specified destination.
+// TODO: Add retry logic.
+func (u *uploader) Upload(ctx context.Context, metaArchive *LedgerMetaArchive) error {
+	logger.Infof("Uploading: %s", metaArchive.GetObjectKey())
+
+	err := u.destination.PutFileIfNotExists(ctx, metaArchive.GetObjectKey(),
+		&XDRGzipEncoder{XdrPayload: &metaArchive.data})
+	if err != nil {
+		return errors.Wrapf(err, "error uploading %s", metaArchive.GetObjectKey())
+	}
+	return nil
+}
+
+// Run starts the uploader, continuously listening for LedgerMetaArchive objects to upload.
+func (u *uploader) Run(ctx context.Context) error {
+	uploadCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-ctx.Done()
+		logger.Info("Context done, waiting for remaining uploads to complete...")
+		// allow up to 10 seconds to upload remaining objects from metaArchiveCh
+		<-time.After(10 * time.Second)
+		logger.Info("Timeout reached, canceling remaining uploads...")
+		cancel()
+	}()
+
+	for {
+		select {
+		case <-uploadCtx.Done():
+			return uploadCtx.Err()
+
+		case metaObject, ok := <-u.metaArchiveCh:
+			if !ok {
+				logger.Info("Meta archive channel closed, stopping uploader")
+				return errors.New("Meta archive channel closed")
```

Comment:
I don't think this should return an error here. Consider the case where we export a fixed number of ledgers, eventually the export manager will iterate through the fixed number and exit without error. Similarly, the uploader should also exit without error.

Even in the case where the export manager encounters an error and then closes the channel. The export manager will return an error. But the uploader should just return nil here because from its perspective there was no error in the upload logic

```suggestion
				return nil
```

Reply:
Yes, I added that while adding unit tests to validate the channel close case. But I noticed it's logged even in the case of a clean shutdown, as you rightly noted. I will remove the error and update the test to validate that it returns cleanly in the case of channel close(as opposed to panicking).